readme.md
# babel-plugin-syntax-highlight

A [Babel](https://babeljs.io) plugin that transforms the code contents of template literals lead by comments specifying a [Prism.js](https://prismjs.com) language into syntax highlighted HTML.

Build-time syntax highlighting advantages:

- 🚀 No runtime JS to slow page loads.
- 🖌 Less client rendering work.
- 🎨 Beautiful code the instant the HTML and CSS loads.

## Installation

To install [`babel-plugin-syntax-highlight`](https://npm.im/babel-plugin-syntax-highlight) with [npm](https://npmjs.com/get-npm), run:

```sh
npm install babel-plugin-syntax-highlight --save-dev
```

Configure [Babel](https://babeljs.io) to use the plugin:

```json
{
  "plugins": ["syntax-highlight"]
}
```

## Examples

In a file transpiled by [Babel](https://babeljs.io), lead a template literal containing syntax to highlight with a comment containing `syntax-highlight` and a [Prism.js language name](https://prismjs.com/#supported-languages):

```js
const highlighted = /* syntax-highlight graphql */ `
  scalar Upload
`;
```

The comment may be surrounded by others on the same or other lines for editor syntax highlighting, [Prettier](https://prettier.io) formatting, etc.

```js
const highlighted =
  /* syntax-highlight graphql */
  /* GraphQL */ `
    scalar Upload
  `;
```

A block or line comment may be used:

```js
const highlighted =
  // syntax-highlight graphql
  `scalar Upload`;
```

The plugin removes the comment (preserving surrounding comments) and transforms the template literal contents into HTML:

```js
const highlighted = `<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;
```

Styling the HTML is up to you; there are many theme stylesheets available. Often themes require `<pre>` and `<code>` containers with appropriate language classes.

A [React](https://reactjs.org) component can be used display highlighted code with appropriate HTML and styles:

```jsx
function SyntaxHighlightedCode({ language, code }) {
  return (
    <pre className={`language-${language}`}>
      <code
        className={`language-${language}`}
        dangerouslySetInnerHTML={{ __html: code }}
      />
    </pre>
  );
}
```

```jsx
const syntaxHighlightedCode = (
  <SyntaxHighlightedCode language="css" code={highlighted} />
);
```

## Requirements

Supported runtime environments:

- [Node.js](https://nodejs.org) versions `^14.17.0 || ^16.0.0 || >= 18.0.0`.

Projects must configure [TypeScript](https://typescriptlang.org) to use types from the modules that have a `// @ts-check` comment:

- [`compilerOptions.allowJs`](https://typescriptlang.org/tsconfig#allowJs) should be `true`.
- [`compilerOptions.maxNodeModuleJsDepth`](https://typescriptlang.org/tsconfig#maxNodeModuleJsDepth) should be reasonably large, e.g. `10`.
- [`compilerOptions.module`](https://typescriptlang.org/tsconfig#module) should be `"node16"` or `"nodenext"`.

## Exports

The [npm](https://npmjs.com) package [`babel-plugin-syntax-highlight`](https://npm.im/babel-plugin-syntax-highlight) features [optimal JavaScript module design](https://jaydenseric.com/blog/optimal-javascript-module-design). These CommonJS modules are exported via the [`package.json`](./package.json) field [`exports`](https://nodejs.org/api/packages.html#exports):

- [`babelPluginSyntaxHighlight.js`](./babelPluginSyntaxHighlight.js)


changelog.md
# babel-plugin-syntax-highlight changelog

## 4.0.0

### Major

- Updated Node.js support to `^14.17.0 || ^16.0.0 || >= 18.0.0`.
- Updated dev dependencies, some of which require newer Node.js versions than previously supported.
- Removed the package `main` field.
- Removed `./package` from the package `exports` field; the full `package.json` filename must be used in a `require` path.
- Renamed the `index.js` module to `babelPluginSyntaxHighlight.js` and added it to the package `exports` field so it can be deep imported.
- Implemented TypeScript types via JSDoc comments.
- Use the `node:` URL scheme for Node.js builtin module imports in tests.

### Patch

- Updated dependencies.
- Simplified dev dependencies and config for ESLint.
- Simplified package scripts.
- Check TypeScript types via a new package `types` script.
- Various type safety improvements.
- Guard against the edge case where the specified Prism.js language name is the name of one of the `Prism.languages` methods, e.g. `extend`.
- Gracefully error if a template literal string can’t be cooked, e.g. due to an invalid escape sequence with the Babel transform option `parserOpts.errorRecovery` enabled.
- Updated GitHub Actions CI config:
  - No longer run tests with Windows.
  - Run tests with Node.js v14, v16, v18.
  - Updated `actions/checkout` to v3.
  - Updated `actions/setup-node` to v3.
- Configured Prettier option `singleQuote` to the default, `false`.
- Updated tests:
  - Use the [`@babel/core`](https://npm.im/@babel/core) function `transformAsync` instead of `transform`.
  - Use a new [`snapshot-assertion`](https://npm.im/snapshot-assertion) dev dependency to snapshot test error messages.
- Added a `license.md` MIT License file.
- Revamped the readme:
  - Removed the badges.
  - Changed headings.
  - Added a “Requirements” section.
  - Added information about TypeScript config and [optimal JavaScript module design](https://jaydenseric.com/blog/optimal-javascript-module-design).
  - Added an “Exports” section.

## 3.0.0

### Major

- Updated Node.js version support to `^12.20 || >= 14.13`.
- Updated dev dependencies, some of which require newer Node.js versions than were previously supported.
- Added a package `exports` field.
- Added a package `sideEffects` field.
- The tests are now ESM in an `.mjs` file instead of CJS in a `.js` file.

### Patch

- Updated dev dependencies.
- Updated GitHub Actions CI config:
  - Also run on pull request.
  - Updated `actions/checkout` to v2.
  - Updated `actions/setup-node` to v2.
  - Run with Node.js v12, v14, v15, v16.
  - Use the simpler [`npm install-test`](https://docs.npmjs.com/cli/v7/commands/npm-install-test) command.
  - Don’t specify the `CI` environment variable as it’s set by default.
- Updated the package scripts:
  - Simpler JSDoc related scripts now that [`jsdoc-md`](https://npm.im/jsdoc-md) v10 automatically generates a Prettier formatted readme.
  - Added a `test:jsdoc` script that checks the readme API docs are up to date with the source JSDoc.
  - Simpler `test:prettier` script arguments.
- Configured Prettier option `semi` to the default, `true`.
- Removed `npm-debug.log` from the `.gitignore` file as npm [v4.2.0](https://github.com/npm/npm/releases/tag/v4.2.0)+ doesn’t create it in the current working directory.
- Updated the EditorConfig.
- Removed the readme “Support” section.
- Added the `.js` file extension to a [`prismjs`](https://npm.im/prismjs) `require` path.
- Always use regex `u` mode.
- Readme tweaks.

## 2.1.0

### Minor

- Setup [GitHub Sponsors funding](https://github.com/sponsors/jaydenseric):
  - Added `.github/funding.yml` to display a sponsor button in GitHub.
  - Added a `package.json` `funding` field to enable npm CLI funding features.

### Patch

- Updated dev dependencies.
- Removed the now redundant [`eslint-plugin-import-order-alphabetical`](https://npm.im/eslint-plugin-import-order-alphabetical) dev dependency.
- Use [`coverage-node`](https://npm.im/coverage-node) for test code coverage.
- Use strict mode for scripts.
- Renamed the `test:units` script to `test:api`.
- Use double quotes around the `test:prettier` script glob for Windows support.
- Added a `.gitattributes` file to enforce `LF` line endings in a Windows environment.
- Test Node.js v13 in CI.
- Prettier ignore `package.json`.
- Comment typo fix.
- Renamed a variable in tests.
- Added more tests.

## 2.0.1

### Patch

- Updated dev dependencies.
- Removed now redundant `eslint-disable-next-line no-console` comments.
- Fixed `${` not being escaped as `\${`.
- Use [`test-director`](https://npm.im/test-director) for tests.
- Moved the readme “Support” section to the end.

## 2.0.0

### Major

- Changed the API from importing a template literal tag, to leading an untagged template literal with a comment that specifies the Prism.js language.
- Removed the now redundant `moduleSpecifier` Babel plugin option.
- Syntax highlighted template literals are now replaced with another template literal, instead of a string.

### Patch

- Updated dev dependencies.
- Updated package description.
- Removed the readme “Todo” section. It’s better to use GitHub issues.
- Aesthetic enhancements to the `TestManager` utility class for managing tests.
- Test that Babel syntax errors have pretty source location info.

## 1.0.0

Initial release.


test.js
'use strict'

const assert = require('assert').strict
const babel = require('@babel/core')
const { TestDirector } = require('test-director')
const babelPluginSyntaxHighlight = require('.')

const tests = new TestDirector()

/**
 * Creates a function that validates an error is a Babel syntax error with a
 * particular message that contains an annotated source location.
 * @param {RegExp} messageRegex RegEx to check the error message against.
 * @returns {Function} Error validator.
 */
const annotatedBabelSyntaxErrorValidator = messageRegex => error =>
  // Error is a Babel syntax error.
  error.name === 'SyntaxError' &&
  // Error has a particular message.
  messageRegex.test(error.message) &&
  // Error message has an annotated source location.
  />.+\d+.+\|.+\^/su.test(error.message)

tests.add('Line comment expression statement.', () => {
  assert.strictEqual(
    babel.transform('// syntax-highlight graphql\n`scalar Upload`', {
      plugins: [babelPluginSyntaxHighlight]
    }).code,
    '`<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  )
})

tests.add('Block comment expression statement.', () => {
  assert.strictEqual(
    babel.transform('/* syntax-highlight graphql */ `scalar Upload`', {
      plugins: [babelPluginSyntaxHighlight]
    }).code,
    '`<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  )
})

tests.add('Variable declarator.', () => {
  assert.strictEqual(
    babel.transform(
      'const a = /* syntax-highlight graphql */ `scalar Upload`',
      { plugins: [babelPluginSyntaxHighlight] }
    ).code,
    'const a = `<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  )
})

tests.add('Arrow function expression.', () => {
  assert.strictEqual(
    babel.transform(
      'const a = () => /* syntax-highlight graphql */ `scalar Upload`',
      { plugins: [babelPluginSyntaxHighlight] }
    ).code,
    'const a = () => `<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  )
})

tests.add('Object property.', () => {
  assert.strictEqual(
    babel.transform(
      'const a = { b: /* syntax-highlight graphql */ `scalar Upload` }',
      { plugins: [babelPluginSyntaxHighlight] }
    ).code,
    'const a = {\n  b: `<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`\n};'
  )
})

tests.add('Template literal escapes.', () => {
  assert.strictEqual(
    babel.transform(
      // Use markup for simple output. Backslash escaping of ‘`’, ‘\’ and ‘${’
      // (only if ‘$’ is followed by ‘{’) is being tested.
      '/* syntax-highlight markup */ `\\`\\${ $ \\\\ \\` \\``',
      { plugins: [babelPluginSyntaxHighlight] }
    ).code,
    '`\\`\\${ $ \\\\ \\` \\``;'
  )
})

tests.add('Unrelated expression statement template literal.', () => {
  assert.strictEqual(
    babel.transform('``', {
      plugins: [babelPluginSyntaxHighlight]
    }).code,
    '``;'
  )
})

tests.add('Unrelated variable declaration template literal.', () => {
  assert.strictEqual(
    babel.transform('const a = ``', {
      plugins: [babelPluginSyntaxHighlight]
    }).code,
    'const a = ``;'
  )
})

tests.add('Unrelated comment and template literal.', () => {
  assert.strictEqual(
    babel.transform('/* GraphQL */ ``', {
      plugins: [babelPluginSyntaxHighlight]
    }).code,
    '/* GraphQL */\n``;'
  )
})

tests.add('Multiple relevant and irrelevant comments.', () => {
  assert.strictEqual(
    babel.transform(
      '/* syntax-highlight graphql */ /* GraphQL */ `scalar Upload`',
      { plugins: [babelPluginSyntaxHighlight] }
    ).code,
    '/* GraphQL */\n`<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  )
})

tests.add('Multiple relevant comments.', () => {
  assert.throws(() => {
    babel.transform(
      '/* syntax-highlight css */ /* syntax-highlight graphql */ `scalar Upload`',
      { plugins: [babelPluginSyntaxHighlight] }
    )
  }, annotatedBabelSyntaxErrorValidator(/Multiple Prism\.js language names specified\./u))
})

tests.add('Comment missing the language name.', () => {
  assert.throws(() => {
    babel.transform('/* syntax-highlight */ ``', {
      plugins: [babelPluginSyntaxHighlight]
    })
  }, annotatedBabelSyntaxErrorValidator(/Missing the Prism\.js language name\./u))
})

tests.add('Unavailable Prism.js language name.', () => {
  assert.throws(() => {
    babel.transform('/* syntax-highlight _ */ ``', {
      plugins: [babelPluginSyntaxHighlight]
    })
  }, annotatedBabelSyntaxErrorValidator(/`_` isn’t an available Prism\.js language name\./u))
})

tests.add('Unsupported template literal expression.', () => {
  assert.throws(() => {
    babel.transform("/* syntax-highlight graphql */ `${''}`", {
      plugins: [babelPluginSyntaxHighlight]
    })
  }, annotatedBabelSyntaxErrorValidator(/Template literal expressions aren’t supported\./u))
})

tests.add('Comment not leading a template literal.', () => {
  assert.strictEqual(
    babel.transform('/* syntax-highlight graphql */', {
      plugins: [babelPluginSyntaxHighlight]
    }).code,
    '/* syntax-highlight graphql */'
  )
})

tests.run()


test.mjs
// @ts-check

import babel from "@babel/core";
import { ok, strictEqual } from "node:assert";
import assertSnapshot from "snapshot-assertion";
import TestDirector from "test-director";

import babelPluginSyntaxHighlight from "./babelPluginSyntaxHighlight.js";

const tests = new TestDirector();

tests.add("Line comment expression statement.", async () => {
  const result = await babel.transformAsync(
    "// syntax-highlight graphql\n`scalar Upload`",
    {
      plugins: [babelPluginSyntaxHighlight],
    }
  );

  strictEqual(
    result?.code,
    '`<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  );
});

tests.add("Block comment expression statement.", async () => {
  const result = await babel.transformAsync(
    "/* syntax-highlight graphql */ `scalar Upload`",
    {
      plugins: [babelPluginSyntaxHighlight],
    }
  );

  strictEqual(
    result?.code,
    '`<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  );
});

tests.add("Variable declarator.", async () => {
  const result = await babel.transformAsync(
    "const a = /* syntax-highlight graphql */ `scalar Upload`",
    { plugins: [babelPluginSyntaxHighlight] }
  );

  strictEqual(
    result?.code,
    'const a = `<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  );
});

tests.add("Arrow function expression.", async () => {
  const result = await babel.transformAsync(
    "const a = () => /* syntax-highlight graphql */ `scalar Upload`",
    { plugins: [babelPluginSyntaxHighlight] }
  );

  strictEqual(
    result?.code,
    'const a = () => `<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  );
});

tests.add("Object property.", async () => {
  const result = await babel.transformAsync(
    "const a = { b: /* syntax-highlight graphql */ `scalar Upload` }",
    { plugins: [babelPluginSyntaxHighlight] }
  );

  strictEqual(
    result?.code,
    'const a = {\n  b: `<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`\n};'
  );
});

tests.add("Template literal escapes.", async () => {
  const result = await babel.transformAsync(
    // Use markup for simple output. Backslash escaping of ‘`’, ‘\’ and ‘${’
    // (only if ‘$’ is followed by ‘{’) is being tested.
    "/* syntax-highlight markup */ `\\`\\${ $ \\\\ \\` \\``",
    { plugins: [babelPluginSyntaxHighlight] }
  );

  strictEqual(result?.code, "`\\`\\${ $ \\\\ \\` \\``;");
});

tests.add("Unrelated expression statement template literal.", async () => {
  const result = await babel.transformAsync("``", {
    plugins: [babelPluginSyntaxHighlight],
  });

  strictEqual(result?.code, "``;");
});

tests.add("Unrelated variable declaration template literal.", async () => {
  const result = await babel.transformAsync("const a = ``", {
    plugins: [babelPluginSyntaxHighlight],
  });

  strictEqual(result?.code, "const a = ``;");
});

tests.add("Unrelated comment and template literal.", async () => {
  const result = await babel.transformAsync("/* GraphQL */ ``", {
    plugins: [babelPluginSyntaxHighlight],
  });

  strictEqual(result?.code, "/* GraphQL */\n``;");
});

tests.add("Multiple relevant and irrelevant comments.", async () => {
  const result = await babel.transformAsync(
    "/* syntax-highlight graphql */ /* GraphQL */ `scalar Upload`",
    { plugins: [babelPluginSyntaxHighlight] }
  );

  strictEqual(
    result?.code,
    '/* GraphQL */\n`<span class="token keyword">scalar</span> <span class="token class-name">Upload</span>`;'
  );
});

tests.add("Multiple relevant comments.", async () => {
  let error;

  try {
    await babel.transformAsync(
      "/* syntax-highlight css */ /* syntax-highlight graphql */ `scalar Upload`",
      { plugins: [babelPluginSyntaxHighlight] }
    );
  } catch (e) {
    error = e;
  }

  ok(error instanceof Error);

  await assertSnapshot(
    error.message,
    new URL(
      "./test/snapshots/babelPluginSyntaxHighlight/multiple-relevant-comments-error.ans",
      import.meta.url
    )
  );
});

tests.add("Comment missing the language name.", async () => {
  let error;

  try {
    await babel.transformAsync("/* syntax-highlight */ ``", {
      plugins: [babelPluginSyntaxHighlight],
    });
  } catch (e) {
    error = e;
  }

  ok(error instanceof Error);

  await assertSnapshot(
    error.message,
    new URL(
      "./test/snapshots/babelPluginSyntaxHighlight/comment-missing-the-language-name-error.ans",
      import.meta.url
    )
  );
});

tests.add(
  "Unavailable Prism.js language name, non `Prism.languages` method name.",
  async () => {
    let error;

    try {
      await babel.transformAsync("/* syntax-highlight _ */ ``", {
        plugins: [babelPluginSyntaxHighlight],
      });
    } catch (e) {
      error = e;
    }

    ok(error instanceof Error);

    await assertSnapshot(
      error.message,
      new URL(
        "./test/snapshots/babelPluginSyntaxHighlight/unavailable-prismjs-language-name-non-prism-languages-method-name-error.ans",
        import.meta.url
      )
    );
  }
);

tests.add(
  "Unavailable Prism.js language name, `Prism.languages` method name.",
  async () => {
    let error;

    try {
      await babel.transformAsync("/* syntax-highlight extend */ ``", {
        plugins: [babelPluginSyntaxHighlight],
      });
    } catch (e) {
      error = e;
    }

    ok(error instanceof Error);

    await assertSnapshot(
      error.message,
      new URL(
        "./test/snapshots/babelPluginSyntaxHighlight/unavailable-prismjs-language-name-prism-languages-method-name-error.ans",
        import.meta.url
      )
    );
  }
);

tests.add("Unsupported template literal expression.", async () => {
  let error;

  try {
    await babel.transformAsync('/* syntax-highlight graphql */ `${""}`', {
      plugins: [babelPluginSyntaxHighlight],
    });
  } catch (e) {
    error = e;
  }

  ok(error instanceof Error);

  await assertSnapshot(
    error.message,
    new URL(
      "./test/snapshots/babelPluginSyntaxHighlight/unsupported-template-literal-expression-error.ans",
      import.meta.url
    )
  );
});

tests.add("Template literal string that can’t be cooked.", async () => {
  let error;

  try {
    await babel.transformAsync(
      "/* syntax-highlight text */ `Invalid escape sequence: \\unicode`",
      {
        plugins: [babelPluginSyntaxHighlight],
        parserOpts: {
          errorRecovery: true,
        },
      }
    );
  } catch (e) {
    error = e;
  }

  ok(error instanceof Error);

  await assertSnapshot(
    error.message,
    new URL(
      "./test/snapshots/babelPluginSyntaxHighlight/template-literal-that-cant-be-cooked-error.ans",
      import.meta.url
    )
  );
});

tests.add("Comment not leading a template literal.", async () => {
  const result = await babel.transformAsync("/* syntax-highlight graphql */", {
    plugins: [babelPluginSyntaxHighlight],
  });

  strictEqual(result?.code, "/* syntax-highlight graphql */");
});

tests.run();


TestManager.js
const chalk = require('chalk')

/** Test manager. */
class TestManager {
  constructor() {
    this.tests = new Map()
  }

  /**
   * Adds a test.
   * @param {string} name Unique test name.
   * @param {Function} test Test to run; may return a Promise.
   */
  addTest(name, test) {
    if (this.tests.has(name))
      throw new Error(`A test called \`${name}\` has already been added.`)

    if (typeof test !== 'function') throw new Error('Test must be a function.')

    this.tests.set(name, test)
  }

  /**
   * Runs all tests.
   * @returns {Promise<void>} Resolves once all tests have run.
   */
  async runTests() {
    let passCount = 0
    let failCount = 0

    for (const [name, test] of this.tests) {
      console.group(`\nTest: ${chalk.bold(name)}`)

      try {
        await test()
        passCount++
      } catch (error) {
        failCount++
        console.error(chalk.dim.red(error.stack))
      } finally {
        console.groupEnd()
      }
    }

    console.info(
      `\n${chalk.bold[failCount ? 'red' : 'green'](
        `${passCount}/${this.tests.size} tests passed.`
      )}\n`
    )

    if (failCount)
      // eslint-disable-next-line no-process-exit
      process.exit(1)
  }
}

module.exports.TestManager = TestManager


babelPluginSyntaxHighlight.js
// @ts-check

"use strict";

const Prism = require("prismjs");
const loadLanguages = require("prismjs/components/index.js");

// Load every available Prism.js language.
loadLanguages();

/**
 * A [Babel](https://babeljs.io) plugin that transforms the code contents of
 * template literals lead by comments specifying a
 * [Prism.js](https://prismjs.com) language into syntax highlighted HTML.
 * @param {import("@babel/core") & import("@babel/core").ConfigAPI} babelApi
 *   Babel API.
 * @returns {import("@babel/core").PluginObj} Babel plugin.
 */
function babelPluginSyntaxHighlight({ types }) {
  return {
    name: "syntax-highlight",
    visitor: {
      TemplateLiteral(
        path,

        // Destructuring the state would break the context for it’s instance
        // methods.
        state
      ) {
        // The language specifier comment is the first match before the template
        // literal, allowing coexistence with other leading comments like
        // /* GraphQL */ for editor syntax highlighting and Prettier formatting.
        // The user may mistakenly use multiple language specifier comments; all
        // must be gathered to be able to alert the user of their mistake.

        /**
         * The leading comments. Sometimes Babel associates them with an
         * ancestor, instead of the template literal.
         */
        const leadingComments = path.node.leadingComments
          ? path.node.leadingComments
          : path.parent.type === "ExpressionStatement"
          ? path.parent.leadingComments
          : null;

        // Skip this template literal if there are no leading comments.
        if (!leadingComments) return;

        /**
         * @type {Array<{
         *   leadingCommentIndex: number,
         *   comment: import("@babel/core").types.Comment,
         *   languageName: string
         * }>}
         */
        const languageSpecifierComments = [];

        leadingComments.forEach((comment, leadingCommentIndex) => {
          // A matching comment begins with optional whitespace (including
          // newlines), followed by “syntax-highlight”, then whitespace
          // (including newlines), then any non-whitespace characters are
          // captured as the Prism.js language name, then optional whitespace
          // characters until the end. Examples:
          // 1. /* syntax-highlight css */
          // 2. // syntax-highlight css
          const match = /^\s*syntax-highlight(?:\s+([^\s]+))?\s*$/mu.exec(
            // A comment value excludes the actual comment syntax.
            comment.value
          );

          if (match) {
            if (!match[1])
              // @ts-ignore Incorrect types.
              throw state.buildCodeFrameError(
                comment,
                "Missing the Prism.js language name."
              );

            languageSpecifierComments.push({
              leadingCommentIndex,
              comment,
              languageName: match[1],
            });
          }
        });

        if (languageSpecifierComments.length) {
          if (languageSpecifierComments.length > 1)
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              languageSpecifierComments[1].comment,
              "Multiple Prism.js language names specified."
            );

          const { languageName } = languageSpecifierComments[0];

          if (
            !(languageName in Prism.languages) ||
            // Guard against the edge case where the specified language name is
            // the name of one of the `Prism.languages` methods, e.g. `extend`.
            typeof Prism.languages[
              /** @type {keyof typeof Prism.languages} */ (languageName)
            ] !== "object"
          )
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              languageSpecifierComments[0].comment,
              `\`${languageName}\` isn’t an available Prism.js language name.`
            );

          const prismGrammar = /** @type {import("prismjs").Grammar} */ (
            Prism.languages[
              /** @type {keyof typeof Prism.languages} */ (languageName)
            ]
          );

          if (path.node.expressions.length)
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              path.node.expressions[0],
              "Template literal expressions aren’t supported."
            );

          /** The code string from the template literal. */
          const text = path.node.quasis[0].value.cooked;

          if (typeof text !== "string")
            // Babel can’t provide the cooked string, possibly due to an invalid
            // escape sequence in the raw string. Normally that results in a
            // Babel parser error, but that won’t happen if the Babel transform
            // option `parserOpts.errorRecovery` is enabled.
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              path.node.quasis[0],
              "Template literal string can’t be cooked."
            );

          /** The syntax highlighted code string. */
          const highlightedCode = Prism.highlight(
            text,
            prismGrammar,
            languageName
          );

          // Remove the language specifier comment. This has to happen before
          // replacing the template literal, or else the triggered revisit will
          // find the comment again and start a crazy feedback loop.
          leadingComments.splice(
            languageSpecifierComments[0].leadingCommentIndex,
            1
          );

          // Replace the template literal with the syntax highlighted version.
          path.replaceWith(
            types.templateLiteral(
              [
                types.templateElement(
                  {
                    raw: highlightedCode
                      // Add a backslash before every ‘`’, ‘\’ and ‘${’.
                      // See: https://github.com/babel/babel/issues/9242#issuecomment-532529613
                      .replace(/\\|`|\$\{/gu, "\\$&"),
                    cooked: highlightedCode,
                  },
                  true
                ),
              ],
              []
            )
          );
        }
      },
    },
  };
}

module.exports = babelPluginSyntaxHighlight;


index.js
// @ts-check

"use strict";

const Prism = require("prismjs");
const loadLanguages = require("prismjs/components/index.js");

// Load every available Prism.js language.
loadLanguages();

/**
 * A [Babel](https://babeljs.io) plugin that transforms the code contents of
 * template literals lead by comments specifying a
 * [Prism.js](https://prismjs.com) language into syntax highlighted HTML.
 * @param {import("@babel/core") & import("@babel/core").ConfigAPI} babelApi
 *   Babel API.
 * @returns {import("@babel/core").PluginObj} Babel plugin.
 */
function babelPluginSyntaxHighlight({ types }) {
  return {
    name: "syntax-highlight",
    visitor: {
      TemplateLiteral(
        path,

        // Destructuring the state would break the context for it’s instance
        // methods.
        state
      ) {
        // The language specifier comment is the first match before the template
        // literal, allowing coexistence with other leading comments like
        // /* GraphQL */ for editor syntax highlighting and Prettier formatting.
        // The user may mistakenly use multiple language specifier comments; all
        // must be gathered to be able to alert the user of their mistake.

        /**
         * The leading comments. Sometimes Babel associates them with an
         * ancestor, instead of the template literal.
         */
        const leadingComments = path.node.leadingComments
          ? path.node.leadingComments
          : path.parent.type === "ExpressionStatement"
          ? path.parent.leadingComments
          : null;

        // Skip this template literal if there are no leading comments.
        if (!leadingComments) return;

        /**
         * @type {Array<{
         *   leadingCommentIndex: number,
         *   comment: import("@babel/core").types.Comment,
         *   languageName: string
         * }>}
         */
        const languageSpecifierComments = [];

        leadingComments.forEach((comment, leadingCommentIndex) => {
          // A matching comment begins with optional whitespace (including
          // newlines), followed by “syntax-highlight”, then whitespace
          // (including newlines), then any non-whitespace characters are
          // captured as the Prism.js language name, then optional whitespace
          // characters until the end. Examples:
          // 1. /* syntax-highlight css */
          // 2. // syntax-highlight css
          const match = /^\s*syntax-highlight(?:\s+([^\s]+))?\s*$/mu.exec(
            // A comment value excludes the actual comment syntax.
            comment.value
          );

          if (match) {
            if (!match[1])
              // @ts-ignore Incorrect types.
              throw state.buildCodeFrameError(
                comment,
                "Missing the Prism.js language name."
              );

            languageSpecifierComments.push({
              leadingCommentIndex,
              comment,
              languageName: match[1],
            });
          }
        });

        if (languageSpecifierComments.length) {
          if (languageSpecifierComments.length > 1)
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              languageSpecifierComments[1].comment,
              "Multiple Prism.js language names specified."
            );

          const { languageName } = languageSpecifierComments[0];

          if (
            !(languageName in Prism.languages) ||
            // Guard against the edge case where the specified language name is
            // the name of one of the `Prism.languages` methods, e.g. `extend`.
            typeof Prism.languages[
              /** @type {keyof typeof Prism.languages} */ (languageName)
            ] !== "object"
          )
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              languageSpecifierComments[0].comment,
              `\`${languageName}\` isn’t an available Prism.js language name.`
            );

          const prismLanguage = /** @type {import("prismjs").Grammar} */ (
            Prism.languages[
              /** @type {keyof typeof Prism.languages} */ (languageName)
            ]
          );

          if (path.node.expressions.length)
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              path.node.expressions[0],
              "Template literal expressions aren’t supported."
            );

          /** The code string from the template literal. */
          const text = path.node.quasis[0].value.cooked;

          if (typeof text !== "string")
            // Babel can’t provide the cooked string, possibly due to an invalid
            // escape sequence in the raw string. Normally that results in a
            // Babel parser error, but that won’t happen if the Babel transform
            // option `parserOpts.errorRecovery` is enabled.
            // @ts-ignore Incorrect types.
            throw state.buildCodeFrameError(
              path.node.quasis[0],
              "Template literal string can’t be cooked."
            );

          /** The syntax highlighted code string. */
          const highlightedCode = Prism.highlight(
            text,
            prismLanguage,
            languageName
          );

          // Remove the language specifier comment. This has to happen before
          // replacing the template literal, or else the triggered revisit will
          // find the comment again and start a crazy feedback loop.
          leadingComments.splice(
            languageSpecifierComments[0].leadingCommentIndex,
            1
          );

          // Replace the template literal with the syntax highlighted version.
          path.replaceWith(
            types.templateLiteral(
              [
                types.templateElement(
                  {
                    raw: highlightedCode
                      // Add a backslash before every ‘`’, ‘\’ and ‘${’.
                      // See: https://github.com/babel/babel/issues/9242#issuecomment-532529613
                      .replace(/\\|`|\$\{/gu, "\\$&"),
                    cooked: highlightedCode,
                  },
                  true
                ),
              ],
              []
            )
          );
        }
      },
    },
  };
}

module.exports = babelPluginSyntaxHighlight;
